cpp/src/arrow/util/async_iterator.h

```diff
@@ -0,0 +1,295 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+#include <queue>
+
+#include "arrow/util/functional.h"
+#include "arrow/util/future.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/optional.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+template <typename T>
+using AsyncGenerator = std::function<Future<T>()>;
+
+/// Iterates through a generator of futures, visiting the result of each one and
+/// returning a future that completes when all have been visited
+template <typename T>
+Future<> VisitAsyncGenerator(AsyncGenerator<T> generator,
+                             std::function<Status(T)> visitor) {
+  auto loop_body = [generator, visitor] {
```

Comment:
Wouldn't that callable still need to be copied?

If they carry much state they are probably un-copyable and they wouldn't have been able to fit into a std::function in the first place.  There are many places where we copy these functions around.  For example, the visitor is immediately copied again in line 39.

I've been operating under the assumption that anything put into a std::function should be cheap to copy.  In reader.cc you'll notice that both transformers are quite big and stateful so I go to the effort of wrapping them in a shared_ptr and capturing that shared_ptr with a lambda.

Also, `VisitAsyncGenerator` is eventually going to expand to `VisitAsyncGenerator(AsyncGenerator<T> generator, std::function<Status(T)> visitor, int parallelism)`.  In that case the visitor is going to be copied up to `parallelism` times so it is important it be copyable.

Reply:
Hmm, I don't think it hurts to minimize copies in any case. Not sure what @bkietz thinks about this?

Also, perhaps the "wrap a callable in a shared_ptr" pattern can be eased with a dedicated primitive?

Reply:
Ok, so I converted it to locally scoped structs.  They are named though...is there some trick to make them fully anonymous or is this what you were thinking?

It removes the copy of the generator.  The copy of the visitor is unavoidable.  It will be added as a callback to multiple futures.  So the visitor will get copied for every future that comes in.

Reply:
> is there some trick to make them fully anonymous or is this what you were thinking?

No, this is what I was thinking (bad terminology, sorry :-)).